File: app/src/main/java/org/breezyweather/main/dialogs/HourlyWeatherDialog.java
package org.breezyweather.main.dialogs;

import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import org.breezyweather.common.basic.models.Location;
import org.breezyweather.common.basic.models.options.unit.PrecipitationUnit;
import org.breezyweather.common.basic.models.options.unit.ProbabilityUnit;
import org.breezyweather.common.basic.models.options.unit.TemperatureUnit;
import org.breezyweather.common.basic.models.weather.Hourly;
import org.breezyweather.common.basic.models.weather.WeatherCode;
import org.breezyweather.common.ui.widgets.AnimatableIconView;
import org.breezyweather.theme.resource.ResourceHelper;
import org.breezyweather.theme.resource.ResourcesProviderFactory;
import org.breezyweather.theme.resource.providers.ResourceProvider;
import org.breezyweather.R;
import org.breezyweather.settings.SettingsManager;

public class HourlyWeatherDialog {

    public static void show(Activity activity, Location location, Hourly hourly) {
        View view = LayoutInflater
                .from(activity)
                .inflate(R.layout.dialog_weather_hourly, null, false);
        initWidget(view, hourly);

        new MaterialAlertDialogBuilder(activity)
                .setTitle(
                        hourly.getHour(activity, location.getTimeZone())
                                + " - "
                                + hourly.getLongDate(activity, location.getTimeZone())
                )
                .setView(view)
                .show();
    }

    private static void initWidget(View view, Hourly hourly) {
        ResourceProvider provider = ResourcesProviderFactory.getNewInstance();

        AnimatableIconView weatherIcon = view.findViewById(R.id.dialog_weather_hourly_icon);

        view.findViewById(R.id.dialog_weather_hourly_weatherContainer).setOnClickListener(v ->
                weatherIcon.startAnimators());

        WeatherCode weatherCode = hourly.getWeatherCode();
        boolean daytime = hourly.isDaylight();
        weatherIcon.setAnimatableIcon(
                ResourceHelper.getWeatherIcons(provider, weatherCode, daytime),
                ResourceHelper.getWeatherAnimators(provider, weatherCode, daytime)
        );

        TextView weatherText = view.findViewById(R.id.dialog_weather_hourly_text);

        SettingsManager settings = SettingsManager.getInstance(view.getContext());
        TemperatureUnit temperatureUnit = settings.getTemperatureUnit();
        PrecipitationUnit precipitationUnit = settings.getPrecipitationUnit();

        StringBuilder builder = new StringBuilder(
                hourly.getWeatherText()
                        + ",  "
                        + hourly.getTemperature().getTemperature(view.getContext(), temperatureUnit)
        );
        if (hourly.getTemperature().getFeelsLikeTemperature() != null) {
            builder.append("\n")
                    .append(view.getContext().getString(R.string.feels_like))
                    .append(" ")
                    .append(hourly.getTemperature().getFeelsLikeTemperature(view.getContext(), temperatureUnit));
        }
        if (hourly.getPrecipitation().getTotal() != null) {
            Float p = hourly.getPrecipitation().getTotal();
            builder.append("\n")
                    .append(view.getContext().getString(R.string.precipitation))
                    .append(" : ")
                    .append(precipitationUnit.getValueText(view.getContext(), p));
        }
        if (hourly.getPrecipitationProbability().getTotal() != null
                && hourly.getPrecipitationProbability().getTotal() > 0) {
            Float p = hourly.getPrecipitationProbability().getTotal();
            builder.append("\n")
                    .append(view.getContext().getString(R.string.precipitation_probability))
                    .append(" : ")
                    .append(ProbabilityUnit.PERCENT.getValueText(view.getContext(), (int) (float) p));
        }
        weatherText.setText(builder.toString());
    }
}

File: app/src/main/java/org/breezyweather/main/dialogs/HourlyWeatherDialog.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.main.dialogs

import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import breezyweather.domain.location.model.Location
import breezyweather.domain.weather.model.Hourly
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import org.breezyweather.R
import org.breezyweather.common.extensions.currentLocale
import org.breezyweather.common.extensions.getFormattedMediumDayAndMonth
import org.breezyweather.common.extensions.getHour
import org.breezyweather.common.ui.widgets.AnimatableIconView
import org.breezyweather.settings.SettingsManager
import org.breezyweather.theme.resource.ResourceHelper
import org.breezyweather.theme.resource.ResourcesProviderFactory
import java.text.NumberFormat

object HourlyWeatherDialog {
    fun show(activity: Activity, location: Location, hourly: Hourly) {
        val view = LayoutInflater
            .from(activity)
            .inflate(R.layout.dialog_weather_hourly, null, false)
        initWidget(view, hourly)
        MaterialAlertDialogBuilder(activity)
            .setTitle(
                hourly.date.getHour(location, activity) +
                    " - " +
                    hourly.date.getFormattedMediumDayAndMonth(location, activity)
            )
            .setView(view)
            .show()
    }

    private fun initWidget(view: View, hourly: Hourly) {
        val provider = ResourcesProviderFactory.newInstance
        val weatherIcon = view.findViewById<AnimatableIconView>(R.id.dialog_weather_hourly_icon)
        view.findViewById<View>(R.id.dialog_weather_hourly_weatherContainer)
            .setOnClickListener { weatherIcon.startAnimators() }
        val weatherCode = hourly.weatherCode
        val daytime = hourly.isDaylight
        if (weatherCode != null) {
            weatherIcon.setAnimatableIcon(
                ResourceHelper.getWeatherIcons(provider, weatherCode, daytime),
                ResourceHelper.getWeatherAnimators(provider, weatherCode, daytime)
            )
        }
        val weatherText = view.findViewById<TextView>(R.id.dialog_weather_hourly_text)
        val settings = SettingsManager.getInstance(view.context)
        val temperatureUnit = settings.temperatureUnit
        val precipitationUnit = settings.precipitationUnit
        val builder = StringBuilder()
        hourly.weatherText?.let {
            builder.append(it)
        }
        hourly.temperature?.temperature?.let {
            if (builder.toString().isNotEmpty()) {
                builder.append(view.context.getString(R.string.comma_separator))
            }
            builder.append(
                temperatureUnit.getValueText(view.context, it)
            )
        }
        hourly.temperature?.feelsLikeTemperature?.let {
            if (builder.toString().isNotEmpty()) builder.append("\n")
            builder.append(view.context.getString(R.string.temperature_feels_like))
                .append(" ")
                .append(
                    temperatureUnit.getValueText(view.context, it)
                )
        }
        hourly.precipitation?.total?.let {
            if (builder.toString().isNotEmpty()) builder.append("\n")
            builder.append(view.context.getString(R.string.precipitation))
                .append(view.context.getString(R.string.colon_separator))
                .append(precipitationUnit.getValueText(view.context, it))
        }
        if ((hourly.precipitationProbability?.total ?: 0.0) > 0) {
            if (builder.toString().isNotEmpty()) builder.append("\n")
            builder.append(view.context.getString(R.string.precipitation_probability))
                .append(view.context.getString(R.string.colon_separator))
                .append(
                    NumberFormat.getPercentInstance(view.context.currentLocale).apply {
                        maximumFractionDigits = 0
                    }.format(hourly.precipitationProbability!!.total!!.div(100.0))
                )
        }
        weatherText.text = builder.toString()
    }
}
